compiler/qsc_fir/src/fir.rs

```diff
@@ -310,6 +313,239 @@ impl Display for Res {
+    pub package: PackageId,
+    /// The item ID.
+    pub item: LocalItemId,
+}
+
+impl Display for StoreItemId {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "<item {} in package {}>", self.item, self.package)
+    }
+}
+
+impl From<(PackageId, LocalItemId)> for StoreItemId {
+    fn from(tuple: (PackageId, LocalItemId)) -> Self {
+        Self {
+            package: tuple.0,
+            item: tuple.1,
+        }
+    }
+}
+
+/// A unique identifier for a block within a package store.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct StoreBlockId {
+    /// The package ID.
+    pub package: PackageId,
+    /// The item ID.
+    pub block: BlockId,
+}
+
+impl Display for StoreBlockId {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "<block {} in package {}>", self.block, self.package)
+    }
+}
+
+impl From<(PackageId, BlockId)> for StoreBlockId {
+    fn from(tuple: (PackageId, BlockId)) -> Self {
+        Self {
+            package: tuple.0,
+            block: tuple.1,
+        }
+    }
+}
+
+/// A unique identifier for an expression within a package store.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct StoreExprId {
+    /// The package ID.
+    pub package: PackageId,
+    /// The expression ID.
+    pub expr: ExprId,
+}
+
+impl Display for StoreExprId {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "<expression {} in package {}>", self.expr, self.package)
+    }
+}
+
+impl From<(PackageId, ExprId)> for StoreExprId {
+    fn from(tuple: (PackageId, ExprId)) -> Self {
+        Self {
+            package: tuple.0,
+            expr: tuple.1,
+        }
+    }
+}
+
+/// A unique identifier for a pattern within a package store.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct StorePatId {
+    /// The package ID.
+    pub package: PackageId,
+    /// The pat ID.
+    pub pat: PatId,
+}
+
+impl Display for StorePatId {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "<pattern {} in package {}>", self.pat, self.package)
+    }
+}
+
+impl From<(PackageId, PatId)> for StorePatId {
+    fn from(tuple: (PackageId, PatId)) -> Self {
+        Self {
+            package: tuple.0,
+            pat: tuple.1,
+        }
+    }
+}
+
+/// A unique identifier for a statement within a package store.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct StoreStmtId {
+    /// The package ID.
+    pub package: PackageId,
+    /// The statement ID.
+    pub stmt: StmtId,
+}
+
+impl Display for StoreStmtId {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "<pattern {} in package {}>", self.stmt, self.package)
+    }
+}
+
+impl From<(PackageId, StmtId)> for StoreStmtId {
+    fn from(tuple: (PackageId, StmtId)) -> Self {
+        Self {
+            package: tuple.0,
+            stmt: tuple.1,
+        }
+    }
+}
+
+/// A trait to find elements in a package store.
+pub trait PackageStoreLookup {
```

Comment:
Does this really need to be a trait if it has only one impl?

Reply:
I think keeping this trait even if it has only one implementation was a few advantages:
- IMO it keeps code more organized because otherwise we would implement these functions in the `fir::PackageStore` implementation.
- When the trait is used in arguments to functions, it makes it clear that the argument is only used as a lookup. The functions that receive a `PackageStoreLookup` argument can't do other things like iterate through packages or get individual packages.

Having said that, I don't have enough experience with Rust to know if this is a common pattern so I'm open to change it if you think these advantages are not enough to warrant a trait with only one implementation.

Reply:
I like it this way, actually. It's funny because I did the exact same thing in a PR where I named a trait `Lookup`, had only one implementation, and got the same question that I just asked you 😄  see: https://github.com/microsoft/qsharp/pull/759/files#r1380776877  . So I completely understand your reasoning, I was just curious what your take was!

Reply:
I liked the God object mention in your response!